### PR TITLE
feat(service_grpc_flight): JDBC compatible Handshake

### PR DESCRIPTION
**Pull request by mhilton**
Closes influxdata/idpe#17522

While testing some JDBC based clients we found that some, Tableau
in this case,  cannot be configured with authoriztion tokens. In
these cases we need to be able to support username/password. The
approach taken is to ignore the username and make the token the
password. This is the same approach being taken throughout the
product.

To facilitate this the Flight RPC Handshake command has been extended
to look for Basic authorization credentials and respond with the
appropriate Bearer authorization header.

While adding end-to-end tests the subprocess commands were causing
a deadlock. These have been changed to using the tonic::process
module.

There are also some small changes to the JDBC test application where
the hardcoded values were clashing with the authorization parameters.
Describe your proposed changes here.

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

**Comment by mhilton** [comment]
@wiedld / @domodwyer I ~~stole~~ shared the authorization header code parsing code from the write path. I'd be greatful if you gave it a look over to make sure I didn't break your hard work.

File: service_grpc_flight/src/lib.rs [hunk 1]
```diff
@@ -527,18 +530,34 @@ where
         &self,
         request: Request<Streaming<HandshakeRequest>>,
     ) -> Result<Response<Self::HandshakeStream>, tonic::Status> {
+        let authz_token = get_flight_authz(request.metadata());
```
Comment (mhilton) [inline service_grpc_flight/src/lib.rs:533]: The JDBC driver doesn't send the `namespace_name` metadata in the handshake request, so we cannot actually do any access checking here.

**Comment by alamb** [comment]
I plan to review this carefully tomorrow

File: service_grpc_flight/src/lib.rs [hunk 2]
```diff
@@ -522,18 +525,34 @@ where
...
+
         let request = request
             .into_inner()
             .message()
             .await?
             .context(InvalidHandshakeSnafu)?;
 
+        let response_header = authz_token
```
Comment (alamb) [inline service_grpc_flight/src/lib.rs:536]:
```suggestion
        // The handshake method is used for authentication. IOx ignores the
        // username and returns the password itself as the token to use for
        // subsequent requests
        let response_header = authz_token
```

File: influxdb_iox/tests/end_to_end_cases/flightsql.rs [hunk 3]
```diff
@@ -1220,113 +1221,257 @@ async fn flightsql_jdbc() {
...
     )
     .run()
-    .await
+    .await;
+
+    authz.close().await;
+}
+
+#[tokio::test]
+/// Runs  the `jdbc_client` program against IOx to verify authenticated JDBC via FlightSQL is working
```
Comment (alamb) [inline influxdb_iox/tests/end_to_end_cases/flightsql.rs:1302]:
```suggestion
/// Runs  the `jdbc_client` program against IOx to verify authenticated JDBC via FlightSQL is working
///
/// In this mode the username is empty and password is the authorization token
```

**Review by alamb — APPROVED** [review]
The code and tests look great to me 👌 👨‍🍳  very nice work @mhilton 

I left some small comments, mostly about adding comments, that I do not think are required to merge this PR. 

One thing that I think would help is to document the Handshake protocol here
https://github.com/influxdata/influxdb_iox/blob/09d6b4ae50ada08aa398d2652e993366b58a0990/service_grpc_flight/src/lib.rs#L257-L258

(aka that the username/password are supplied and then the password is returned as the bearer token for subsequent requests)

Mostly I think the documentation would allow us to have a single place to point at if anyone wanted to know how the protocol worked in IOx

File: authz/src/lib.rs [hunk 4]
```diff
@@ -41,14 +67,14 @@ pub trait Authorizer: std::fmt::Debug + Send + Sync {
     /// empty permission sets.
     async fn permissions(
         &self,
-        token: Option<&[u8]>,
+        token: Option<Vec<u8>>,
```
Comment (domodwyer) [inline authz/src/lib.rs:70]: Nice - thank you :+1:

**Review by domodwyer — APPROVED** [review]
Thanks for the drive-by elimination of the unnecessary clone :+1:

**Pull request merged**